### PR TITLE
Implement move-all.

### PR DESCRIPTION
**Pull request by gbggrant**
Modifed arrays move and cram move.
Updated tests and integration tests.

### Description

_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any JIRA issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: Ensure that you have added or updated tests
- [ ] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

### CODE REVIEW

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala
Comment (gbggrant) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: This is a wholesale copy from the base class - probably could be done a lot better, but couldn't quickly find a way to only get the pathsForDelivery without filtering them out.

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala
Comment (gbggrant) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala]: Need to modify the Cram delivery executor to not deliver the fingerprints (which I don't think it does anyway?).
In any case, I think the comment is inappropriate here as we *can* move fingerprints

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala
Comment (danxmoran) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: Rough idea: add a `deliverInto` method to the [`DeliverableMetadata` trait](https://github.com/broadinstitute/clio/blob/develop/clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/DeliverableMetadata.scala), and call it here.

File: clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/ArraysExtensions.scala
Comment (pshapiro4broad) [inline clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/ArraysExtensions.scala:16]: It looks like this is only used in tests. If so, it should be defined in a test class.

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala
Comment (pshapiro4broad) [inline clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala]: Not sure what the point of this change is, was this required to get the tests to pass?

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/CramTests.scala
Comment (pshapiro4broad) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/CramTests.scala]: What is the status of these tests? If they are no longer valid, the should be removed.

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala
Comment (gbggrant) [reply clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala]: It's the actual extension of green idat files, so I thought it more useful that the code handle those.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: This `override` should get dropped, and the existing `override` in the `DeliverExecutor` should be changed to not make its `super` call.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala
Comment (danxmoran) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: Walking myself back after time away, a few things:
1. This is going to do more than just move files, so I wouldn't have `Move` in its name
2. The copied logic is set up to avoid too much boilerplate when the concrete type of `moveCommand.index.MetadataType` isn't known. For the delivery executors, we know exactly what the type of the metadata is, so there's nothing stopping us from calling `copy` on the case class / manually building the list of `IoOps`. I think it'd make the code a lot clearer to do the simple thing here (and in the cram delivery executor, when you get to it).

Comment (danxmoran) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: For example, see how the existing `buildDelivery` explicitly accesses the idat paths. Now that I'm looking at it I think all of this logic should go into that method.

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala
Comment (danxmoran) [reply clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala]: Cram delivery only moves the cram + crai, and generates the md5. It should stay that way (at least in this PR).

File: clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala
Comment (danxmoran) [inline clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala]: Reminder: there are more metrics files to fill in here before merging. I think everything `URI` in this class should be getting moved as part of move execution.

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala
Comment (danxmoran) [reply clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala]: You'll need to change the logic in the `DeliverArraysExecutor` which currently uses the generic idat extension for this to have an effect on main code.

File: clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala
Comment (danxmoran) [reply clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala]: We could probably come up with some type-fu to extract out stuff-to-move based on field types... not sure if it's worth it.

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala
Comment (gbggrant) [reply clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala]: check

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/CramTests.scala
Comment (gbggrant) [reply clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/CramTests.scala]: removed now.  Superseded by the testing above

File: clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/ArraysExtensions.scala
Comment (gbggrant) [reply clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/arrays/ArraysExtensions.scala:16]: It's used in DeliverArraysExecutor now.

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala
Comment (gbggrant) [reply clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala]: check.

File: clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala
Comment (gbggrant) [reply clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala]: There are some that are redundant (fingerprintPath vs fingerprintVcfPath) which I guess I can move both of,
and there are some that are probably legacy (preBqsrDepthSmPath) that I can't find in recent workflows and don't know the extension for.

**Comment by danxmoran** [comment]
Confirming question from Slack: ripping out all of the expectations [like this](https://github.com/broadinstitute/clio/blob/develop/clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/MoveExecutorSpec.scala#L110) from the `MoveExecutorSpec` should fix most of the current failures. [This test](https://github.com/broadinstitute/clio/blob/develop/clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/MoveExecutorSpec.scala#L267) will probably need to be deleted entirely. Not sure what's up with the failure in the `RetrieveAndPrintExecutor`... will look for anything fishy in the code diff.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: I'd call this `existingMetadata` to be clear the method is transforming pre-delivery state into post-delivery state.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: This filter would be worth defining once in the `MoveExecutor`, in a separate follow-up method. We'd need to make it handle all 3 cases (but that should be OK). Both moves and copies get dropped if src == dest, and writes always happen(?).

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: Shouldn't need the `baseStream` / `flatMapConcat` wrapper any more; could just call `buildDelivery(metadata.withWorkspaceName(...))`

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/CramTests.scala
Comment (danxmoran) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/CramTests.scala]: I feel bad asking to add more, but I'd also want to check:
1. All readgroup-level metrics (stored as an array in Cram metadata) get moved
2. All logs (pointed to by a top-level directory) get moved

Should be possible to test with 4 extra files.

File: clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala
Comment (danxmoran) [inline clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala]: Test I asked for will make sure this gets uncommented.

I think you want something like:
```scala
readgroupLevelMetricsFiles = readgroupLevelMetricsFiles.map(pathMapper(_, ???))
```
Though as I typed it I realized there's no consistent extension to use...

**Review by danxmoran — COMMENTED** [review]
Nearly there!

I'm not sure what to do about the readgroup-level metrics files. We could punt on it for now and make a separate ticket specifically for dealing with them.

File: clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala
Comment (danxmoran) [reply clio-transfer-model/src/main/scala/org/broadinstitute/clio/transfer/model/wgscram/CramMetadata.scala]: Something like this might work:
```scala
readgroupLevelMetricsFiles = readgroupLevelMetricsFiles.map { path =>
  val name = new File(path).getName
  pathMapper(path, name)
}
```
It's a bad hack, but I think it'll fool the existing code. Better would be for us to restructure this logic to distinguish things-which-can-change-their-base-name from things-which-can-only-be-moved, but that's for another ticket.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala
Comment (gbggrant) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: check

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala
Comment (gbggrant) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: Thanks.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala
Comment (gbggrant) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: check.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/MoveExecutor.scala
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/MoveExecutor.scala]: Slightly less code: `case MoveOp(src, dest) => src.equals(dest)`, same for the copy case.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/MetadataMover.scala
Comment (pshapiro4broad) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/MetadataMover.scala:51]: Does this comment still belong here? There is no `newBasename` and there's no code that handles file extensions here.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/MetadataMover.scala
Comment (pshapiro4broad) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/MetadataMover.scala:28]: scaladoc please

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala
Comment (pshapiro4broad) [inline clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala:87]: it looks like the negative tests weren't ported, was there a reason for that?

File: clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala
Comment (pshapiro4broad) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala:26]: only `logPath` doesn't require handling `newBasename`?

File: clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala
Comment (pshapiro4broad) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala:108]: it looks like this will overwrite `fingerprintVcfPath` since it has the same file extension

File: clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala
Comment (pshapiro4broad) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala:245]: no renaming required here?

**Review by pshapiro4broad — APPROVED** [review]
Looks ok, too bad about all the code expansion

**Review by danxmoran — APPROVED** [review]
Needs a scalafmt, but LGTM after

File: clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala
Comment (danxmoran) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala:108]: The fields refer to the same file. Looks like we don't actually use `fingerprintVcfPath` in prod based on querying:
```
{
    "query": {
        "exists": { "field": "fingerprint_vcf_path" }
    }
}
```
And getting no results. We could take the chance to remove the field now.

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala
Comment (danxmoran) [reply clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutorSpec.scala:87]: Checking these cases got dropped from the main code

File: clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala
Comment (danxmoran) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala:26]: `logPath` and `readgroupLevelMetrics`. But `logPath` doesn't actually work as-is, so this should get removed for now.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala
Comment (danxmoran) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/metadata/CramMover.scala:245]: Nope. If we did rename we'd get naming conflicts all over the place (unless we did something like `newBase + full old name`)

**Pull request merged**
